stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,787 @@
+  static constexpr bool is_var_t0_{is_var<T_t0>::value};
+  static constexpr bool is_var_y0_{is_var<T_y0_t0>::value};
+  static constexpr bool is_any_var_args_{
+      disjunction<is_var<scalar_type_t<T_Args>>...>::value};
+  static constexpr bool is_var_return_{is_var<T_Return>::value};
+  static constexpr bool is_var_only_ts_{
+      is_var_ts_ && !(is_var_t0_ || is_var_y0_ || is_any_var_args_)};
+
+  arena_t<Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1>> y0_;
+  arena_t<T_t0> t0_;
+  std::vector<arena_t<T_ts>, arena_allocator<arena_t<T_ts>>> ts_;
+
+  double relative_tolerance_forward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_forward_;
+  double relative_tolerance_backward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_backward_;
+  double relative_tolerance_quadrature_;
+  double absolute_tolerance_quadrature_;
+  long int max_num_steps_;                  // NOLINT(runtime/int)
+  long int num_steps_between_checkpoints_;  // NOLINT(runtime/int)
+  int interpolation_polynomial_;
+  int solver_forward_;
+  int solver_backward_;
+
+  std::ostream* msgs_;
+
+  std::tuple<arena_t<
+      plain_type_t<decltype(deep_copy_vars(std::declval<const T_Args&>()))>>...>
+      local_args_tuple_;
+  std::tuple<arena_t<
+      plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>>...>
+      value_of_args_tuple_;
+
+  size_t N_;
+  bool backward_is_initialized_;
+
+  size_t num_args_vars_;
+
+  arena_t<Eigen::VectorXd> state_forward_;
+  arena_t<Eigen::VectorXd> state_backward_;
+  arena_t<Eigen::VectorXd> quad_;
+
+  vari** non_chaining_varis_;
+
+  vari** args_varis_;
+
+  int index_backward_;
+
+  /**
+   * Since the CVODES solver manages memory with malloc calls, these resources
+   * must be freed using a destructor call (which is not being called for the
+   * vari class).
+   */
+  struct cvodes_solver : public chainable_alloc {
+    const std::decay_t<F> f_;
+    const size_t N_;
+    std::vector<Eigen::VectorXd> y_;
+    std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> y_return_;
+    const std::string function_name_str_;
+    const char* function_name_;
+    void* cvodes_mem_;
+    N_Vector nv_state_forward_;
+    N_Vector nv_state_backward_;
+    N_Vector nv_quad_;
+    N_Vector nv_absolute_tolerance_forward_;
+    N_Vector nv_absolute_tolerance_backward_;
+
+    SUNMatrix A_forward_;
+    SUNLinearSolver LS_forward_;
+
+    SUNMatrix A_backward_;
+    SUNLinearSolver LS_backward_;
+
+    template <typename FF, typename StateFwd, typename StateBwd, typename Quad,
+              typename AbsTolFwd, typename AbsTolBwd>
+    cvodes_solver(const char* function_name, FF&& f, size_t N,
+                  size_t num_args_vars, size_t ts_size, int solver_forward,
+                  StateFwd& state_forward, StateBwd& state_backward, Quad& quad,
+                  AbsTolFwd& absolute_tolerance_forward,
+                  AbsTolBwd& absolute_tolerance_backward)
+        : chainable_alloc(),
```

Comment:
Mostly I just think it's kosher to call it first. Though yeah I think C++ calls the inherited constructor automatically

Reply:
I'd prefer calling the base class constructor as well to make it explicit. In case the constructor would take arguments, then we also would have to call it.